src/common/inference/bm25.rs

```diff
@@ -0,0 +1,163 @@
+use std::collections::{BTreeMap, HashMap};
+use std::sync::Arc;
+
+use collection::operations::point_ops::VectorPersisted;
+use itertools::Itertools;
+use murmur3::murmur3_32_of_slice;
+use segment::data_types::index::{StemmingAlgorithm, StopwordsInterface, TokenizerType};
+use segment::index::field_index::full_text_index::stop_words::StopwordsFilter;
+use segment::index::field_index::full_text_index::tokenizers::{
+    Stemmer, Tokenizer, TokenizerConfig,
+};
+use serde::{Deserialize, Serialize};
+
+/// Configuration of the local bm25 models.
+#[derive(Debug, Deserialize, Serialize, Clone)]
+pub struct Bm25Config {
+    #[serde(default = "default_k")]
+    pub k: f64,
+    #[serde(default = "default_b")]
+    pub b: f64,
+    #[serde(default = "default_avg_len")]
+    pub avg_len: f64,
+    #[serde(default)]
+    pub tokenizer: TokenizerType,
+    #[serde(default, flatten)]
+    pub text_preprocessing_config: Option<TextPreprocessingConfig>,
+}
+
+/// Bm25 tokenizer configurations.
+#[derive(Debug, Deserialize, Serialize, Clone, Default)]
+pub struct TextPreprocessingConfig {
+    #[serde(default)]
+    pub lowercase: bool,
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub stopwords_filter: Option<StopwordsInterface>,
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub stemmer: Option<StemmingAlgorithm>,
+    pub min_token_len: Option<usize>,
+    pub max_token_len: Option<usize>,
+}
+
+const fn default_k() -> f64 {
+    1.2
+}
+
+const fn default_b() -> f64 {
+    0.75
+}
+
+const fn default_avg_len() -> f64 {
+    256.0
+}
+
+impl Default for Bm25Config {
+    fn default() -> Self {
+        Self {
+            k: default_k(),
+            b: default_b(),
+            avg_len: default_avg_len(),
+            tokenizer: Default::default(),
+            text_preprocessing_config: Default::default(),
+        }
+    }
+}
+
+/// Bm25 implementation
+#[derive(Debug)]
+pub struct Bm25 {
+    config: Bm25Config,
+    tokenizer: Tokenizer,
+}
+
+impl Bm25 {
+    pub fn new(mut config: Bm25Config) -> Self {
+        let tokenizer_conf = std::mem::take(&mut config.text_preprocessing_config);
+        let tokenizer_config = TokenizerConfig::from(tokenizer_conf.unwrap_or_default());
+        let tokenizer = Tokenizer::new(config.tokenizer, tokenizer_config);
+        Self { config, tokenizer }
+    }
+
+    /// Tokenizes the `input` with the configured tokenizer options.
+    fn tokenize<'b>(&self, input: &'b str) -> Vec<Cow<'b, str>> {
+        let mut out = vec![];
+        self.tokenizer.tokenize_query(input, |i| out.push(i));
+        out
+    }
+
+    /// Embedds the given input using the Bm25 algorithm and configured options/hyperparameters.
+    pub fn search_embed(&self, input: &str) -> VectorPersisted {
+        let tokens = self.tokenize(input);
+
+        if tokens.is_empty() {
+            return VectorPersisted::empty_sparse();
+        }
+
+        let indices: Vec<u32> = tokens
```

Comment:
`DimId`?

Reply:
We don't have access to `DimId` and `DimWeight` in `src/common`. I don't think we should add a dependency, just for this type alias.
If you prefer having some sort of alias here, I can add a new type alias here, but I think the semantics is already known from the variable names and context from the function.

The same applies to the other place in line 103